parsec/core/fs/workspacefs/workspacefs_timestamped.py

```diff
@@ -41,16 +45,25 @@ def __init__(self, workspacefs: WorkspaceFS, timestamp: pendulum.Pendulum):
         self.workspace_id = workspacefs.workspace_id
         self.get_workspace_entry = workspacefs.get_workspace_entry
         self.device = workspacefs.device
-        self.local_storage = workspacefs.local_storage
+        self.local_storage = workspacefs.local_storage.to_timestamped(timestamp)
         self.backend_cmds = workspacefs.backend_cmds
         self.event_bus = workspacefs.event_bus
         self.remote_device_manager = workspacefs.remote_device_manager
 
         self.timestamp = timestamp
 
-        self.remote_loader = workspacefs.remote_loader
-        self.file_transactions = workspacefs.file_transactions
-        self.entry_transactions = workspacefs.entry_transactions
+        self.remote_loader = workspacefs.remote_loader.to_timestamped(timestamp)
+        self.file_transactions = FileTransactions(
+            self.workspace_id, self.local_storage, self.remote_loader, self.event_bus
+        )
+        self.entry_transactions = EntryTransactions(
+            self.workspace_id,
+            self.get_workspace_entry,
```

Comment:
You could also mock `get_workspace_entry().role` to `WorkspaceRole.READER` to make sure no write operations is performed (instead of using `_throw_permission_error`)